infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServerCache.java

```diff
@@ -0,0 +1,204 @@
+                .maximumSize(cacheSize)
+                .writer(writer)
+                //Performing periodic maintenance using current thread (synchronously)
+                .executor(Runnable::run)
+                .recordStats()
+                .build();
+    }
+
+    /**
+     * The cache limited by size.
+     * For a synchronous cache we are using a same-thread executor (Runnable::run)
+     * https://github.com/ben-manes/caffeine/issues/90
+     *
+     * @param cacheSize cache size
+     */
+    public SequencerServerCache(long cacheSize) {
+        this.conflictCache = Caffeine.newBuilder()
+                .maximumSize(cacheSize)
+                //Performing periodic maintenance using current thread (synchronously)
+                .executor(Runnable::run)
+                .writer(getDefaultCacheWriter())
+                .recordStats()
+                .build();
+    }
+
+    private CacheWriter<ConflictTxStream, Long> getDefaultCacheWriter() {
+        return new CacheWriter<ConflictTxStream, Long>() {
+            /**
+             * Let caffeine do its work
+             * @param key a key
+             * @param value a value
+             */
+            @Override
+            public void write(@Nonnull ConflictTxStream key, @Nonnull Long value) {
+                //ignore
+            }
+
+            /**
+             * Eviction policy https://github.com/ben-manes/caffeine/wiki/Writer
+             * @param key  conflict stream
+             * @param globalAddress global address
+             * @param cause a removal cause
+             */
+            @Override
+            public void delete(@Nonnull ConflictTxStream key, Long globalAddress, @Nonnull RemovalCause cause) {
+                if (cause == RemovalCause.REPLACED) {
+                    String errMsg = String.format("Override error. Conflict key: %s, address: %s", key, globalAddress);
+                    throw new IllegalStateException(errMsg);
+                }
+
+                log.trace(
+                        "Updating maxConflictWildcard. Old = '{}', new ='{}' conflictParam = '{}'. Cause = '{}'",
+                        maxConflictWildcard, globalAddress, key, cause
+                );
+
+                if (globalAddress == null) {
+                    globalAddress = Address.NOT_FOUND;
+                }
+                maxConflictWildcard = Math.max(globalAddress, maxConflictWildcard);
+            }
+        };
+    }
+
+    /**
+     * Returns the value associated with the {@code key} in this cache,
+     * or {@code null} if there is no cached value for the {@code key}.
+     *
+     * @param conflictKey conflict stream
+     * @return global address
+     */
+    public Long getIfPresent(ConflictTxStream conflictKey) {
+        return conflictCache.getIfPresent(conflictKey);
+    }
+
+    /**
+     * Invalidate all records up to a trim mark.
+     *
+     * @param trimMark trim mark
+     */
+    public void invalidate(long trimMark) {
```

Comment:
I agree with Maithem. Please rename this method to invalidateUpTo. From the name it seems, it's invalidating only one entry. This method invokes `conflictCache.invalidate(key)` which invalidates only 1 key. It can get confusing.

Reply:
done